src/elements/components/FeatheryAuthGate.tsx

```diff
@@ -0,0 +1,178 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { Form } from '../..';
+import { Props as FormProps } from '../../Form';
+import { getStytchJwt } from '../../utils/browser';
+import { getAuthClient, initInfo, initState } from '../../utils/init';
+import { authHookCb } from '../../integrations/stytch';
+import Client from '../../utils/client';
+import { isAuthStytch } from '../../integrations/utils';
+import Spinner from './Spinner';
+import { isHrefFirebaseMagicLink } from '../../integrations/firebase';
+/** TODO: These next 2 should maybe be dynamically imported, but having trouble with that
+ * combined 6.9k gzipped, so OK for now
+ */
+import { useIdleTimer } from 'react-idle-timer';
+import throttle from 'lodash.throttle';
+
+const TEN_SECONDS_IN_MILLISECONDS = 1000 * 10;
+const FIVE_MINUTES_IN_MILLISECONDS = 1000 * 60 * 5;
+
+export const authState = {
+  // This is a flag so we only redirect to the login start step immediately
+  // after auth, not during other form navigation
+  redirectAfterLogin: false,
+  sentAuth: false,
+  clearLoader: () => {},
+  onLogin: () => {},
+  onLogout: () => {}
+};
+
+const FeatheryAuthGate = ({
+  authId: authIdProp,
+  formProps,
+  whitelistPath,
+  fullPageLogin,
+  onLogin = () => {},
+  onLogout = () => {},
+  children
+}: {
+  authId?: string;
+  formProps: FormProps;
+  whitelistPath?: string;
+  fullPageLogin?: boolean;
+  onLogin?: () => void;
+  onLogout?: () => void;
+  children: JSX.Element;
+}) => {
+  const formCompleted = initInfo().sessions[formProps.formName]?.form_completed;
+
+  // Need to use this flag because when doing magic link login the onChange
+  // event doesn't seem to be added early enough to catch the first event which
+  // is the one containing the token. subsequent events do not contain the token
+  const hasAuthedRef = useRef(false);
+  // Use this render state to force re-evaluation of initState, since initState isn't reactive as-is
+  const [render, setRender] = useState(true);
+  const [showLoader, setShowLoader] = useState(false);
+  authHookCb.cb = () => setRender((prev) => !prev);
+
+  const logoutActions = () => {
+    hasAuthedRef.current = false;
+    initState.authId = undefined;
+    onLogout();
+  };
+
+  useEffect(() => {
+    if (
+      // We should set loader for new auth sessions
+      window.location.search.includes('stytch_token_type') ||
+      isHrefFirebaseMagicLink() ||
+      // and existing ones
+      getStytchJwt()
+    ) {
+      authState.redirectAfterLogin = true;
+      setShowLoader(true);
+    }
+
+    const { location, history } = window;
+    if (whitelistPath && location.pathname !== whitelistPath) {
+      // If user is not at the URL whitelisted for auth, take them there for login
+      history.replaceState(null, '', whitelistPath);
+    }
+
+    authState.clearLoader = () => setShowLoader(false);
```

Comment:
do we need this callback function? can't we just hide loader when authState.onLogin is called? aka

authState.onLogin = () => {setShowLoader(false); onLogin()}

Reply:
will double check if there's a reason that doesn't work, or if I was just trying to clear the loader in the same spot as the Form. agreed that would be better